lib/pacemaker/pcmk_sched_colocation.c

```diff
@@ -1337,19 +1337,36 @@ add_node_scores_matching_attr(GHashTable *nodes, const pe_resource_t *rsc,
         int weight = 0;
         int score = 0;
         int new_score = 0;
+        const char *value = pe_node_attribute_raw(node,
+                                                  colocation->node_attribute);
 
-        score = best_node_score_matching_attr(rsc, attr,
-                                              pe_node_attribute_raw(node, attr));
+        score = best_node_score_matching_attr(rsc, colocation->node_attribute,
+                                              value);
 
         if ((factor < 0) && (score < 0)) {
-            /* Negative preference for a node with a negative score
-             * should not become a positive preference.
+            /* If the dependent is anti-colocated, we generally don't want the
+             * primary to prefer nodes that the dependent avoids. That could
+             * lead to unnecessary shuffling of the primary when the dependent
+             * hits its migration threshold somewhere, for example.
              *
-             * @TODO Consider filtering only if weight is -INFINITY
+             * However, there are cases where it is desirable. If the dependent
+             * can't run anywhere but where the primary is, it would be
+             * worthwhile to move the primary for the sake of keeping the
+             * dependent active.
+             *
+             * We use the primary's stickiness to decide what to do, so the user
+             * has some say in the matter. (This is the configured primary, not
+             * a particular instance of the primary, but that doesn't matter
+             * unless stickiness uses a rule to vary by node, and that seems
+             * acceptable to ignore.)
              */
-            crm_trace("%s: Filtering %d + %f * %d (double negative disallowed)",
-                      pe__node_name(node), node->weight, factor, score);
-            continue;
+            if ((colocation == NULL)
```

Comment:
If `colocation == NULL` then we've already seg faulted :) I believe that's also the cause of the coverity error.

Reply:
Whoops